Fallback to RAILS_RELATIVE_URL_ROOT in `url_for`

yasyf opened the pull request:
Fixed an issue where the `RAILS_RELATIVE_URL_ROOT` environment
variable is not prepended to the path when `url_for` is called.
If `SCRIPT_NAME` (used by Rack) is set, it takes precedence.

Fixes #5122.

yasyf commented:
Just a note that I consulted with @pixeltrix before deciding on the order of precedence with these different prefix options.

eileencodes commented on actionpack/test/controller/url_for_test.rb:
```diff
@@ -255,6 +255,16 @@ def test_relative_url_root_is_respected
+        ActionController::Base.config.relative_url_root = '/subdir'
+        add_host!
+        assert_equal('https://www.basecamphq.com/subdir/c/a/i',
+          W.new.url_for(:controller => 'c', :action => 'a', :id => 'i', :protocol => 'https')
```
New tests should use the Ruby 1.9 hash syntax instead of hash rockets.

yasyf commented:
@eileencodes fixed.

rafaelfranca commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -691,7 +691,7 @@ def optimize_routes_generation?
 
       def find_script_name(options)
-        options.delete(:script_name) || ''
+        options.delete(:script_name) || ActionController::Base.config.relative_url_root || ''
```
I'm not sure if we should be coupling Action Dispatch with Action Controller here. Do we have any precedent? cc @pixeltrix 

Can we inject this config at load time using the railtie?

yasyf replied:
I can change this to use railties.

yasyf commented:
@rafaelfranca I added a new config for Action Dispatch to prevent coupling it with Action Controller.

yasyf committed:
Fallback to RAILS_RELATIVE_URL_ROOT in `url_for`.

Fixed an issue where the `RAILS_RELATIVE_URL_ROOT` environment
variable is not prepended to the path when `url_for` is called.
If `SCRIPT_NAME` (used by Rack) is set, it takes precedence.

rafaelfranca commented:
Merged at 44ff0313c121f528a68b3bd21d6c7a96f313e3d3

rafaelfranca commented:
Backported at 0703453fabc4a47411c2f9df291acfdfbcaf5ae4
